### DISCUSSION TITLE
System tray icon

### DISCUSSION THREAD

**Discussion by wl2776** [sourcegit-scm/sourcegit#805] [Q&A]
Is there any way to make SourceGit running permanently and hide into system tray on closing main window?

When I close SourceGit's main window, I'd like to have it running, just hide into the system tray.

**Comment by wl2776** [comment]
I've quickly checked in a fork: system tray icon works in KUbuntu 24.04.

However, I'm not very familiar with C# and .Net, so my development of this feature will be very slow.

**Comment by love-linger** [comment]
I'm sorry. Currently, there're no plans to add this feature.

**Comment by wl2776** [comment]
I've started implementing this feature in https://github.com/wl2776/sourcegit/tree/feature/system-tray-icon

Could you point me to a documentation or tutorials? I cannot figure out how to make main window to unminimize.

I right-click on system tray icon, then click on "Preferences" menu. However, main window does not appear.

![image](https://github.com/user-attachments/assets/2eb8a4fb-3e01-460d-970e-296e208f3ee2)

If I open it manually, it shows preferences dialog.

**Comment by love-linger** [comment]
I haven't seen such a Git GUI Client that offers this. And I really do NOT want to add it to SourceGit.

**Reply by wl2776** [reply]
Please, check SmartGit.
This option is in its low-level properties.

![image](https://github.com/user-attachments/assets/6af77a72-70e8-4d4d-ba90-0acb1d8cd750)

![image](https://github.com/user-attachments/assets/63f7072d-be1b-45b1-b46d-ed5d67890534)


>  And I really do NOT want to add it to SourceGit.

This could be turned off by default, and switched on by editing config file. I will figure out how to do it.

**Comment by heartacker** [comment]
如果可以通过配置，进行开启这个功能，默认关闭，也不是不行。

**Reply by love-linger** [reply]
* SourceGit 仅是一个简单的Git客户端，用于展示当前仓库的状态。并没有启动后台服务的需求。
* SourceGit 提供了Workspace与自动恢复上次打开的仓库功能。
* SourceGit 对于打开仓库的速度进行了大量的优化。
* 该功能在大多数的Git客户端中并未提供
* 在Unix-like平台上，一直保持SourceGit打开会引起其他问题，比如已经提到的FilesystemWatcher相关的问题：#411

---

* `SourceGit` is just a GUI frontend of git that used to display the current status of repository. There're no  background servers started by it.
* We have already provided `Restore tabs on startup` in `Workspace`
* We have done lots of works to improve the speed to open a repository.
* This feature is not provided by most of other git GUI clients too.
* Keeping `SourceGit` running in background may cause issue like #411

**Reply by wl2776** [reply]
> * We have done lots of works to improve the speed to open a repository.

Application launch takes about 3-4 seconds on my Omen HP notebook wih AMD Ryzen 9.
Hiding into system tray and restoring from it takes less than one second.

**Reply by wl2776** [reply]
> * `SourceGit` is just a GUI frontend of git that used to display the current status of repository. There're no  background servers started by it.

My PR does not add any background services, it just changes processing of the `Closing` event in `src/Views/Launcher.axaml.cs`:

```cs
        protected override void OnClosing(WindowClosingEventArgs e)
        {
            var pref = ViewModels.Preference.Instance;
            if (pref.SystemTrayIcon) {
                e.Cancel = true;
                Hide();
            } else {
                (DataContext as ViewModels.Launcher)?.Quit(Width, Height);
            }
            base.OnClosing(e);
        }
```

**Reply by wl2776** [reply]
> * Keeping `SourceGit` running in background may cause issue like [The configured user limit (128) on the number of inotify instances has been reached... #411](https://github.com/sourcegit-scm/sourcegit/issues/411)

So, currently it is not recommended to leave SourceGit running, right?

**Reply by love-linger** [reply]
I know. What I mean is that it's not necessary to keep this app running background

**Reply by wl2776** [reply]
> I know. What I mean is that it's not necessary to keep this app running background

But what if I *want* to keep it running?

For me, personally, the answer is known :)  

I believe, I'm not the only one.

**Comment by love-linger** [comment]
I don't understand that. Do you want to keep the system file manager running background？It is very similar to this app - just a simple GUI frontend of file system.

**Reply by wl2776** [reply]
I always have running a terminal application and a file manager (Krusader and Far2l), very often I launch several copies of them.

My Linux is configured to launch [Yakuake](https://apps.kde.org/yakuake/) on startup.

**Reply by wl2776** [reply]
You can see Yakuake icon on one of my earlier screenshots.

![image](https://github.com/user-attachments/assets/4a897ec1-cdc1-45d7-9a99-5afe359b1563)

**Reply by wl2776** [reply]
> Do you want to keep the system file manager running background？

Please, take a look at Double Commander. 

It is a file manager with the system tray icon.
![image](https://github.com/user-attachments/assets/f05677cf-f556-4ce4-8c7f-ccb3684f83d9)

**Comment by wl2776** [comment]
I've found the issue.
If SourceGit is already running, then launching SourceGit from the OS launcher creates one more system tray icon.

This is not what I'd like to see.

Does SourceGit have anything like single instance mode?

**Comment by wl2776** [comment]
Could anyone check my [PR #868](https://github.com/sourcegit-scm/sourcegit/pull/868) on Macos?
